numba/types/function.py

```diff
@@ -0,0 +1,116 @@
+
+__all__ = ['FunctionType', 'FunctionProtoType', 'WrapperAddressProtocol']
+
+from .abstract import Type, Literal
+
+
+class FunctionType(Type):
+    """The front-end for first-class function types.
+    """
+
+    mutable = True
+    cconv = None
+
+    def __init__(self, ftype):
+        if isinstance(ftype, tuple) and len(ftype) == 2:
+            ftype = FunctionProtoType(ftype[0], ftype[1])
+        if isinstance(ftype, FunctionProtoType):
+            self.ftype = ftype
+            name = ftype.name
+        else:
+            raise TypeError(
+                f'constructing {type(self).__name__} from {type(ftype)}'
+                ' instance is not supported')
+        super(FunctionType, self).__init__(name)
+
+    def signature(self):
+        from numba import typing
+        ptype = self.ftype
+        return typing.signature(ptype.rtype, *ptype.atypes)
+
+    @property
+    def key(self):
+        return self.name
+
+    def cast_python_value(self, value):
+        from numba import typing
+        if isinstance(value, typing.Signature):
+            ptype = FunctionProtoType(value.return_type, value.args)
+            return FunctionType(ptype)
+        raise NotImplementedError(
+            'cast_python_value({}, {})'.format(value, type(value)))
+
+    def get_call_type(self, context, args, kws):
+        from numba import typing
+        ptype = self.ftype
+        if len(args) == len(ptype.atypes):
+            for i, atype in enumerate(args):
+                if isinstance(atype, Literal):
+                    atype = atype.literal_type
+                if not (atype <= ptype.atypes[i]):
+                    break
```

Comment:
wondering if Optional needs handling?

Reply:
yes, likely, and there are other Numba types that could be supported here.
The currently supported Numba types are scalars, literals, arrays.

Reply:
It turns out that supporting `Optional` requires patching numba elsewhere:
```
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of OptionalType(float64(float64)) i.e. the type 'float64(float64) or None' with parameters (float64)
No type info available for OptionalType(float64(float64)) as a callable.
[1] During: resolving callee type: OptionalType(float64(float64)) i.e. the type 'float64(float64) or None'
```
that cannot be fixed by handling Optional here. I suggest to leave it as TODO.